File: src/main/java/org/trp/shincolle/server/FormationService.java
package org.trp.shincolle.server;

import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import org.trp.shincolle.attachment.AdmiralData;
import org.trp.shincolle.entity.base.EntityShipBase;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

public final class FormationService {
    private static final double NEARBY_SHIP_SYNC_RADIUS = 100.0D;
    private static final double NEARBY_SELECTED_IMPORT_RADIUS = 64.0D;

    private FormationService() {
    }

    public static void clearFormationState(EntityShipBase ship) {
        if (ship == null) {
            return;
        }
        ship.setFormationTeam(-1);
        ship.setFormationSlot(-1);
        ship.setPointerSelected(false);
        ship.clearPointerTarget();
        ship.clearPointerTargetEntity();
    }

    public static void applyFormationState(EntityShipBase ship, int teamId, int slotId, boolean selected) {
        if (ship == null) {
            return;
        }
        ship.setFormationTeam(teamId);
        ship.setFormationSlot(slotId);
        ship.setPointerSelected(selected);
    }

    public static void syncNearbyShipsForCurrentTeam(Player player, boolean clearDeselectedTargets) {
        if (player == null) {
            return;
        }
        AdmiralData data = PlayerStateService.admiralData(player);
        int teamId = data.getCurrentTeamID();
        List<EntityShipBase> ships = player.level().getEntitiesOfClass(EntityShipBase.class,
                player.getBoundingBox().inflate(NEARBY_SHIP_SYNC_RADIUS),
                ship -> ship.isOwnedBy(player) && !ship.isInDeadPose());

        for (EntityShipBase ship : ships) {
            if (ship.getFormationTeam() == teamId) {
                int slot = ship.getFormationSlot();
                ship.setPointerSelected(data.isSelected(teamId, slot));
            } else {
                ship.setPointerSelected(false);
                if (clearDeselectedTargets) {
                    ship.clearPointerTarget();
                    ship.clearPointerTargetEntity();
                }
            }
        }
    }

    public static void handleFormationAction(Player player, int action, int param1, int param2,
                                             String paramString, Optional<UUID> paramUuid) {
        if (player == null) {
            return;
        }
        AdmiralData data = PlayerStateService.admiralData(player);

        switch (action) {
            case 0 -> {
                if (PlayerStateService.setCurrentTeamId(player, param1)) {
                    syncNearbyShipsForCurrentTeam(player, true);
                }
            }
            case 1 -> PlayerStateService.setCurrentTeamFormation(player, param1);
            case 2 -> setCurrentTeamSlotSelected(player, data, param1, param2 != 0);
            case 3 -> removeCurrentTeamSlot(player, data, param1);
            case 4 -> PlayerStateService.setCurrentTeamName(player, paramString);
            case 5 -> paramUuid.ifPresent(uuid -> replaceCurrentTeamSlot(player, param1, uuid));
            case 6 -> swapCurrentTeamSlots(player, data, param1, param2);
            case 7 -> importNearbySelectedShips(player, data);
            case 8 -> openCurrentTeamSlotShipMenu(player, data, param1);
            default -> {
            }
        }

        if (player instanceof ServerPlayer serverPlayer) {
            PlayerStateService.sendAdmiralState(serverPlayer);
        }
    }

    public static void handlePointerRosterToggle(Player player, UUID targetUuid) {
        if (player == null || targetUuid == null) {
            return;
        }
        AdmiralData data = PlayerStateService.admiralData(player);
        int teamId = data.getCurrentTeamID();
        int slot = data.findShipSlot(teamId, targetUuid);

        if (slot != -1) {
            boolean nextState = !data.isSelected(teamId, slot);
            if (PlayerStateService.setCurrentTeamSlotSelected(player, slot, nextState)) {
                withServerShip(player, targetUuid, ship -> ship.setPointerSelected(nextState));
            }
        } else {
            int assignedSlot = PlayerStateService.assignShipToCurrentTeam(player, targetUuid);
            if (assignedSlot == -1) {
                player.displayClientMessage(Component.translatable("chat.shincolle.formation.teamfull"), false);
            } else {
                withServerShip(player, targetUuid, ship -> applyFormationState(ship, teamId, assignedSlot, true));
            }
        }

        if (player instanceof ServerPlayer serverPlayer) {
            PlayerStateService.sendAdmiralState(serverPlayer);
        }
    }

    private static void setCurrentTeamSlotSelected(Player player, AdmiralData data, int slotId, boolean selected) {
        if (!PlayerStateService.setCurrentTeamSlotSelected(player, slotId, selected)) {
            return;
        }

        UUID shipUuid = data.getShipUUID(data.getCurrentTeamID(), slotId);
        if (shipUuid != null) {
            withServerShip(player, shipUuid, ship -> ship.setPointerSelected(selected));
        }
    }

    private static void openCurrentTeamSlotShipMenu(Player player, AdmiralData data, int slotId) {
        if (slotId < 0 || slotId >= AdmiralData.SLOT_COUNT) {
            return;
        }

        UUID shipUuid = data.getShipUUID(data.getCurrentTeamID(), slotId);
        if (shipUuid != null) {
            withServerShip(player, shipUuid, ship -> ship.openShipMenu(player));
        }
    }

    private static void removeCurrentTeamSlot(Player player, AdmiralData data, int slotId) {
        if (slotId < 0 || slotId >= AdmiralData.SLOT_COUNT) {
            return;
        }

        UUID shipUuid = data.getShipUUID(data.getCurrentTeamID(), slotId);
        if (shipUuid != null) {
            withServerShip(player, shipUuid, FormationService::clearFormationState);
        }
        PlayerStateService.removeShipFromTeams(player, shipUuid);
    }

    private static void replaceCurrentTeamSlot(Player player, int slotId, UUID shipUuid) {
        PlayerStateService.SlotAssignment assignment = PlayerStateService.setCurrentTeamSlot(player, slotId, shipUuid);
        if (assignment == null) {
            return;
        }

        UUID replacedUuid = assignment.replacedUuid();
        if (replacedUuid != null && !replacedUuid.equals(shipUuid)) {
            withServerShip(player, replacedUuid, FormationService::clearFormationState);
        }
        withServerShip(player, shipUuid, ship -> applyFormationState(ship, assignment.teamId(), assignment.slotId(), true));
    }

    private static void swapCurrentTeamSlots(Player player, AdmiralData data, int slot1, int slot2) {
        if (!PlayerStateService.swapCurrentTeamSlots(player, slot1, slot2)) {
            return;
        }

        int teamId = data.getCurrentTeamID();
        UUID uuid1 = data.getShipUUID(teamId, slot1);
        if (uuid1 != null) {
            withServerShip(player, uuid1, ship -> {
                ship.setFormationSlot(slot1);
                ship.setPointerSelected(data.isSelected(teamId, slot1));
            });
        }

        UUID uuid2 = data.getShipUUID(teamId, slot2);
        if (uuid2 != null) {
            withServerShip(player, uuid2, ship -> {
                ship.setFormationSlot(slot2);
                ship.setPointerSelected(data.isSelected(teamId, slot2));
            });
        }
    }

    private static void importNearbySelectedShips(Player player, AdmiralData data) {
        if (!(player.level() instanceof ServerLevel serverLevel)) {
            return;
        }

        int teamId = data.getCurrentTeamID();
        boolean teamFilledDuringSync = false;
        List<EntityShipBase> nearbySelected = serverLevel.getEntitiesOfClass(EntityShipBase.class,
                player.getBoundingBox().inflate(NEARBY_SELECTED_IMPORT_RADIUS),
                ship -> ship.isPointerSelected() && player.getUUID().equals(ship.getOwnerUUID()));
        List<EntityShipBase> nearbyOwned = serverLevel.getEntitiesOfClass(EntityShipBase.class,
                player.getBoundingBox().inflate(NEARBY_SHIP_SYNC_RADIUS),
                ship -> player.getUUID().equals(ship.getOwnerUUID()) && !ship.isInDeadPose());

        for (EntityShipBase ship : nearbySelected) {
            if (!data.isShipInTeam(teamId, ship.getUUID())
                    && PlayerStateService.assignShipToCurrentTeam(player, ship.getUUID()) == -1) {
                teamFilledDuringSync = true;
            }
        }

        for (EntityShipBase ship : nearbyOwned) {
            if (ship.getFormationTeam() == teamId && !data.isShipInTeam(teamId, ship.getUUID())) {
                clearFormationState(ship);
            }
        }

        for (int i = 0; i < AdmiralData.SLOT_COUNT; i++) {
            UUID uuid = data.getShipUUID(teamId, i);
            if (uuid != null) {
                final int slotId = i;
                withServerShip(player, uuid, ship -> applyFormationState(ship, teamId, slotId, data.isSelected(teamId, slotId)));
            }
        }

        if (teamFilledDuringSync) {
            player.displayClientMessage(Component.translatable("chat.shincolle.formation.teamfull"), false);
        }
    }

    private static void withServerShip(Player player, UUID shipUuid, java.util.function.Consumer<EntityShipBase> action) {
        if (!(player.level() instanceof ServerLevel serverLevel) || shipUuid == null) {
            return;
        }

        Entity entity = serverLevel.getEntity(shipUuid);
        if (entity instanceof EntityShipBase ship) {
            action.accept(ship);
        }
    }
}

File: src/main/java/org/trp/shincolle/server/FormationService.kt
package org.trp.shincolle.server

import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import org.trp.shincolle.attachment.AdmiralData
import org.trp.shincolle.entity.base.EntityShipBase
import java.util.*
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Predicate

object FormationService {
    private const val NEARBY_SHIP_SYNC_RADIUS = 100.0
    private const val NEARBY_SELECTED_IMPORT_RADIUS = 64.0

    fun clearFormationState(ship: EntityShipBase?) {
        if (ship == null) {
            return
        }
        ship.formationTeam = -1
        ship.formationSlot = -1
        ship.isPointerSelected = false
        ship.clearPointerTarget()
        ship.clearPointerTargetEntity()
    }

    fun applyFormationState(ship: EntityShipBase?, teamId: Int, slotId: Int, selected: Boolean) {
        if (ship == null) {
            return
        }
        ship.formationTeam = teamId
        ship.formationSlot = slotId
        ship.isPointerSelected = selected
    }

    fun syncNearbyShipsForCurrentTeam(player: Player?, clearDeselectedTargets: Boolean) {
        if (player == null) {
            return
        }
        val data = PlayerStateService.admiralData(player)
        val teamId = data.getCurrentTeamID()
        val ships = player.level().getEntitiesOfClass<EntityShipBase?>(
            EntityShipBase::class.java,
            player.getBoundingBox().inflate(NEARBY_SHIP_SYNC_RADIUS),
            Predicate { ship: EntityShipBase? -> ship!!.isOwnedBy(player) && !ship.isInDeadPose })

        for (ship in ships) {
            if (ship.formationTeam == teamId) {
                val slot = ship.formationSlot
                ship.isPointerSelected = data.isSelected(teamId, slot)
            } else {
                ship.isPointerSelected = false
                if (clearDeselectedTargets) {
                    ship.clearPointerTarget()
                    ship.clearPointerTargetEntity()
                }
            }
        }
    }

    fun handleFormationAction(
        player: Player?, action: Int, param1: Int, param2: Int,
        paramString: String?, paramUuid: Optional<UUID?>
    ) {
        if (player == null) {
            return
        }
        val data = PlayerStateService.admiralData(player)
        var shouldSync = false

        when (action) {
            0 -> {
                if (PlayerStateService.setCurrentTeamId(player, param1)) {
                    syncNearbyShipsForCurrentTeam(player, true)
                    shouldSync = true
                }
            }

            1 -> shouldSync = PlayerStateService.setCurrentTeamFormation(player, param1)
            2 -> shouldSync = setCurrentTeamSlotSelected(player, data, param1, param2 != 0)
            3 -> shouldSync = removeCurrentTeamSlot(player, data, param1)
            4 -> shouldSync = PlayerStateService.setCurrentTeamName(player, paramString)
            5 -> shouldSync =
                paramUuid.map<Boolean?>(Function { uuid: UUID? -> replaceCurrentTeamSlot(player, param1, uuid) })
                    .orElse(false)

            6 -> shouldSync = swapCurrentTeamSlots(player, data, param1, param2)
            7 -> shouldSync = importNearbySelectedShips(player, data)
            8 -> openCurrentTeamSlotShipMenu(player, data, param1)
            else -> {}
        }

        if (shouldSync && player is ServerPlayer) {
            PlayerStateService.sendAdmiralState(player)
        }
    }

    fun handlePointerRosterToggle(player: Player?, targetUuid: UUID?) {
        if (player == null || targetUuid == null) {
            return
        }
        val data = PlayerStateService.admiralData(player)
        val teamId = data.getCurrentTeamID()
        val slot = data.findShipSlot(teamId, targetUuid)
        var shouldSync = false

        if (slot != -1) {
            val nextState = !data.isSelected(teamId, slot)
            if (PlayerStateService.setCurrentTeamSlotSelected(player, slot, nextState)) {
                withServerShip(
                    player,
                    targetUuid,
                    Consumer { ship: EntityShipBase? -> ship!!.isPointerSelected = nextState })
                shouldSync = true
            }
        } else {
            val assignedSlot = PlayerStateService.assignShipToCurrentTeam(player, targetUuid)
            if (assignedSlot == -1) {
                player.displayClientMessage(Component.translatable("chat.shincolle.formation.teamfull"), false)
            } else {
                withServerShip(
                    player,
                    targetUuid,
                    Consumer { ship: EntityShipBase? -> applyFormationState(ship, teamId, assignedSlot, true) })
                shouldSync = true
            }
        }

        if (shouldSync && player is ServerPlayer) {
            PlayerStateService.sendAdmiralState(player)
        }
    }

    private fun setCurrentTeamSlotSelected(
        player: Player?,
        data: AdmiralData,
        slotId: Int,
        selected: Boolean
    ): Boolean {
        if (!PlayerStateService.setCurrentTeamSlotSelected(player, slotId, selected)) {
            return false
        }

        val shipUuid = data.getShipUUID(data.getCurrentTeamID(), slotId)
        if (shipUuid != null) {
            FormationService.withServerShip(
                player!!,
                shipUuid,
                Consumer { ship: EntityShipBase? -> ship!!.isPointerSelected = selected })
        }
        return true
    }

    private fun openCurrentTeamSlotShipMenu(player: Player, data: AdmiralData, slotId: Int) {
        if (slotId < 0 || slotId >= AdmiralData.SLOT_COUNT) {
            return
        }

        val shipUuid = data.getShipUUID(data.getCurrentTeamID(), slotId)
        if (shipUuid != null) {
            withServerShip(player, shipUuid, Consumer { ship: EntityShipBase? -> ship!!.openShipMenu(player) })
        }
    }

    private fun removeCurrentTeamSlot(player: Player, data: AdmiralData, slotId: Int): Boolean {
        if (slotId < 0 || slotId >= AdmiralData.SLOT_COUNT) {
            return false
        }

        val shipUuid = data.getShipUUID(data.getCurrentTeamID(), slotId)
        if (shipUuid == null) {
            return false
        }
        withServerShip(player, shipUuid, Consumer { obj: EntityShipBase? -> FormationService.clearFormationState(obj) })
        return PlayerStateService.removeShipFromTeams(player, shipUuid)
    }

    private fun replaceCurrentTeamSlot(player: Player?, slotId: Int, shipUuid: UUID?): Boolean {
        val assignment = PlayerStateService.setCurrentTeamSlot(player, slotId, shipUuid)
        if (assignment == null) {
            return false
        }

        val replacedUuid = assignment.replacedUuid
        if (replacedUuid != null && replacedUuid != shipUuid) {
            FormationService.withServerShip(
                player!!,
                replacedUuid,
                Consumer { obj: EntityShipBase? -> FormationService.clearFormationState(obj) })
        }
        FormationService.withServerShip(
            player!!,
            shipUuid,
            Consumer { ship: EntityShipBase? -> applyFormationState(ship, assignment.teamId, assignment.slotId, true) })
        return true
    }

    private fun swapCurrentTeamSlots(player: Player?, data: AdmiralData, slot1: Int, slot2: Int): Boolean {
        if (!PlayerStateService.swapCurrentTeamSlots(player, slot1, slot2)) {
            return false
        }

        val teamId = data.getCurrentTeamID()
        val uuid1 = data.getShipUUID(teamId, slot1)
        if (uuid1 != null) {
            FormationService.withServerShip(player!!, uuid1, Consumer { ship: EntityShipBase? ->
                ship!!.formationSlot = slot1
                ship.isPointerSelected = data.isSelected(teamId, slot1)
            })
        }

        val uuid2 = data.getShipUUID(teamId, slot2)
        if (uuid2 != null) {
            FormationService.withServerShip(player!!, uuid2, Consumer { ship: EntityShipBase? ->
                ship!!.formationSlot = slot2
                ship.isPointerSelected = data.isSelected(teamId, slot2)
            })
        }
        return true
    }

    private fun importNearbySelectedShips(player: Player, data: AdmiralData): Boolean {
        if (player.level() !is ServerLevel) {
            return false
        }

        val teamId = data.getCurrentTeamID()
        var teamFilledDuringSync = false
        var changed = false
        val serverLevel = player.level() as ServerLevel
        val nearbySelected: MutableList<EntityShipBase> = serverLevel.getEntitiesOfClass<EntityShipBase?>(
            EntityShipBase::class.java,
            player.getBoundingBox().inflate(NEARBY_SELECTED_IMPORT_RADIUS),
            Predicate { ship: EntityShipBase? -> ship!!.isPointerSelected && player.getUUID() == ship.ownerUUID })
        val nearbyOwned: MutableList<EntityShipBase> = serverLevel.getEntitiesOfClass<EntityShipBase?>(
            EntityShipBase::class.java,
            player.getBoundingBox().inflate(NEARBY_SHIP_SYNC_RADIUS),
            Predicate { ship: EntityShipBase? -> player.getUUID() == ship!!.ownerUUID && !ship.isInDeadPose })

        for (ship in nearbySelected) {
            if (!data.isShipInTeam(teamId, ship.getUUID())) {
                if (PlayerStateService.assignShipToCurrentTeam(player, ship.getUUID()) == -1) {
                    teamFilledDuringSync = true
                } else {
                    changed = true
                }
            }
        }

        for (ship in nearbyOwned) {
            if (ship.formationTeam == teamId && !data.isShipInTeam(teamId, ship.getUUID())) {
                clearFormationState(ship)
                changed = true
            }
        }

        for (i in 0..<AdmiralData.SLOT_COUNT) {
            val uuid = data.getShipUUID(teamId, i)
            if (uuid != null) {
                val slotId = i
                withServerShip(
                    player,
                    uuid,
                    Consumer { ship: EntityShipBase? ->
                        applyFormationState(
                            ship,
                            teamId,
                            slotId,
                            data.isSelected(teamId, slotId)
                        )
                    })
            }
        }

        if (teamFilledDuringSync) {
            player.displayClientMessage(Component.translatable("chat.shincolle.formation.teamfull"), false)
        }
        return changed
    }

    private fun withServerShip(player: Player, shipUuid: UUID?, action: Consumer<EntityShipBase?>) {
        if (player.level() !is ServerLevel || shipUuid == null) {
            return
        }

        val serverLevel = player.level() as ServerLevel
        val entity: Entity? = serverLevel.getEntity(shipUuid)
        if (entity is EntityShipBase && entity.isOwnedBy(player) && entity.isAlive && !entity.isRemoved) {
            action.accept(entity)
        }
    }
}
